### ISSUE TITLE
FIPS compliance umbrella issue

### ISSUE THREAD

**Issue by qmuntal** [microsoft/go#277]
This is the tracking issue for FIPS work which must be done before saying we are FIPS compliant. The list will grow over time and some of the bullets will have their own issue.

- [x] Several crypto APIs are not backed by boring/openssl because they are not part of the FIPS spec. Decide if we want to panic or not if someone tries to use them.
  - [crypto/des](https://pkg.go.dev/crypto/des): The package documentation states that `DES is cryptographically broken and should not be used for secure applications.`
  - [crypto/md5](https://pkg.go.dev/crypto/md5): The package documentation states that `MD5 is cryptographically broken and should not be used for secure applications.`
  - [ed25519](https://pkg.go.dev/crypto/ed25519): ED5519 is cryptographically secure but not yet approved by FIPS 140-2.

- [x] Several Go crypto packages are not backed by boring/openssl because they don't implement a high level crypto algorithm but are just entities/helpers to be used with other crypto packages. Decide if we want to panic or not if someone tries to use them.
  - [crypto/rand](https://pkg.go.dev/crypto/rand)
  - [crypto/x509](https://pkg.go.dev/crypto/x509)
  - [crypto/elliptic](https://pkg.go.dev/crypto/elliptic)
  - [crypto/cipher](https://pkg.go.dev/crypto/cipher)

- [x] Several FIPS 140-2 crypo algorithms are still using Go crypto instead of boring/openssl. We should open issues upstream or start a discussion in golang-nuts.
  -  [sha512.Sum512_224](https://pkg.go.dev/crypto/sha512#New512_224) and [sha512.Sum512_256](https://pkg.go.dev/crypto/sha512#New512_256): Both supported by openssl 1.1.1 and higher. Should we panic if using 1.0.2 and 1.1.0?

- [x] New crypto APIs: #280 

- [x] Don't force `tlsfips` but make it opt-in
